cloud/linode/route_controller.go

```diff
@@ -0,0 +1,224 @@
+package linode
+
+import (
+	"context"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/linode/linodego"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	cloudprovider "k8s.io/cloud-provider"
+	"k8s.io/klog/v2"
+)
+
+type routeCache struct {
+	sync.RWMutex
+	routes     map[int]*[]linodego.InstanceConfig
+	lastUpdate time.Time
+	ttl        time.Duration
+}
+
+func (rc *routeCache) refreshRoutes(ctx context.Context, client Client) error {
+	rc.Lock()
+	defer rc.Unlock()
+
+	if time.Since(rc.lastUpdate) < rc.ttl {
+		return nil
+	}
+
+	instances, err := client.ListInstances(ctx, nil)
+	if err != nil {
+		return err
+	}
+	rc.routes = make(map[int]*[]linodego.InstanceConfig)
+	for _, instance := range instances {
+		configs, err := client.ListInstanceConfigs(ctx, instance.ID, &linodego.ListOptions{})
+		if err != nil {
+			return err
+		}
+		rc.routes[instance.ID] = &configs
+	}
+	rc.lastUpdate = time.Now()
+
+	return nil
+}
+
+type routes struct {
+	vpcid      int
+	client     Client
+	instances  *instances
+	routeCache *routeCache
+}
+
+func newRoutes(client Client) (cloudprovider.Routes, error) {
+	timeout := 60
+	if raw, ok := os.LookupEnv("LINODE_ROUTES_CACHE_TTL"); ok {
+		if t, _ := strconv.Atoi(raw); t > 0 {
+			timeout = t
+		}
+	}
+
+	vpcid, err := getVPCID(client, Options.VPCName)
+	if err != nil {
+		return nil, err
+	}
+
+	return &routes{
```

Comment:
could we build this route object before we return it? I suspect it will make debugging slightly easier if we can inspect the object before it is returned

Reply:
Done